File: src/main/java/org/frawa/elmtest/run/ElmTestAutoTestManager.java
package org.frawa.elmtest.run;

import com.intellij.execution.testframework.autotest.AbstractAutoTestManager;
import com.intellij.execution.testframework.autotest.AutoTestWatcher;
import com.intellij.execution.testframework.autotest.DelayedDocumentWatcher;
import com.intellij.openapi.components.ServiceManager;
import com.intellij.openapi.components.State;
import com.intellij.openapi.components.Storage;
import com.intellij.openapi.components.StoragePathMacros;
import com.intellij.openapi.fileEditor.FileEditorManager;
import com.intellij.openapi.project.Project;
import org.elm.lang.core.ElmFileType;
import org.jetbrains.annotations.NotNull;

@State(
        name = "ElmTestAutoTestManager",
        storages = {@Storage(StoragePathMacros.WORKSPACE_FILE)}
)
public class ElmTestAutoTestManager extends AbstractAutoTestManager {

    @NotNull
    static ElmTestAutoTestManager getInstance(Project project) {
        return ServiceManager.getService(project, ElmTestAutoTestManager.class);
    }

    ElmTestAutoTestManager(@NotNull Project project) {
        super(project);
    }

    @Override
    @NotNull
    protected AutoTestWatcher createWatcher(Project project) {
        return new DelayedDocumentWatcher(project,
                myDelayMillis,
                this::restartAllAutoTests,
                file -> FileEditorManager.getInstance(project).isFileOpen(file) &&
                        ElmFileType.INSTANCE.equals(file.getFileType())
        );
    }

}

File: src/main/kotlin/org/elm/ide/test/run/ElmTestAutoTestManager.kt
package org.elm.ide.test.run

import com.intellij.execution.testframework.autotest.AbstractAutoTestManager
import com.intellij.execution.testframework.autotest.DelayedDocumentWatcher
import com.intellij.openapi.components.*
import com.intellij.openapi.fileEditor.FileEditorManager
import com.intellij.openapi.project.Project
import com.intellij.openapi.util.Condition
import com.intellij.openapi.vfs.VirtualFile
import com.intellij.util.Consumer
import org.elm.lang.core.ElmFileType

@State(
    name = "ElmTestAutoTestManager",
    storages = [Storage(StoragePathMacros.WORKSPACE_FILE)]
)
@Service(Service.Level.PROJECT)
class ElmTestAutoTestManager internal constructor(
    project: Project
) : AbstractAutoTestManager(project) {

    override fun createWatcher(project: Project) =
        // This is the only constructor that is available both in Platform versions 2022.2.4 and master
        DelayedDocumentWatcher(project,
            myDelayMillis,
            Consumer { value: Int -> restartAllAutoTests(value) },
            Condition { it: VirtualFile -> it.fileType == ElmFileType && FileEditorManager.getInstance(project).isFileOpen(it) }
        )
}

val Project.elmAutoTestManager
    get() = service<ElmTestAutoTestManager>()
